File: server/internal/model/expression_data_test.go
/*
SPDX-FileCopyrightText: 2026 M5Stack Technology CO LTD
SPDX-License-Identifier: MIT
*/

package model

import (
	"encoding/json"
	"testing"
)

func TestExpressionDataJSONRoundTrip(t *testing.T) {
	t.Parallel()

	original := ExpressionData{
		Type: "bleAvatar",
		LeftEye: ExpressionItem{
			X:        1,
			Y:        2,
			Rotation: 3,
			Weight:   4,
			Size:     5,
		},
		RightEye: ExpressionItem{
			X:        6,
			Y:        7,
			Rotation: 8,
			Weight:   9,
			Size:     10,
		},
		Mouth: ExpressionItem{
			X:        11,
			Y:        12,
			Rotation: 13,
			Weight:   14,
			Size:     15,
		},
	}

	data, err := json.Marshal(original)
	if err != nil {
		t.Fatalf("marshal expression data: %v", err)
	}

	var decoded ExpressionData
	if err := json.Unmarshal(data, &decoded); err != nil {
		t.Fatalf("unmarshal expression data: %v", err)
	}

	if decoded != original {
		t.Fatalf("round-trip mismatch: got %+v want %+v", decoded, original)
	}
}

func TestMotionDataJSONRoundTrip(t *testing.T) {
	t.Parallel()

	original := MotionData{
		Type: "bleMotion",
		PitchServo: MotionDataItem{
			Angle: 450,
			Speed: 500,
		},
		YawServo: MotionDataItem{
			Rotate: 120,
			Speed:  300,
		},
	}

	data, err := json.Marshal(original)
	if err != nil {
		t.Fatalf("marshal motion data: %v", err)
	}

	var decoded MotionData
	if err := json.Unmarshal(data, &decoded); err != nil {
		t.Fatalf("unmarshal motion data: %v", err)
	}

	if decoded != original {
		t.Fatalf("round-trip mismatch: got %+v want %+v", decoded, original)
	}
}

func TestDanceDataJSONRoundTrip(t *testing.T) {
	t.Parallel()

	original := DanceData{
		LeftEye: ExpressionItem{Weight: 100},
		RightEye: ExpressionItem{
			Weight: 100,
		},
		Mouth:      ExpressionItem{Size: 25},
		PitchServo: MotionDataItem{Angle: 450, Speed: 900},
		YawServo:   MotionDataItem{Angle: 100, Speed: 700},
		DurationMs: 650,
	}

	data, err := json.Marshal(original)
	if err != nil {
		t.Fatalf("marshal dance data: %v", err)
	}

	var decoded DanceData
	if err := json.Unmarshal(data, &decoded); err != nil {
		t.Fatalf("unmarshal dance data: %v", err)
	}

	if decoded != original {
		t.Fatalf("round-trip mismatch: got %+v want %+v", decoded, original)
	}
}

File: app/test/model/expression_data_test.dart
/*
SPDX-FileCopyrightText: 2026 M5Stack Technology CO LTD
SPDX-License-Identifier: MIT
*/

import 'package:flutter_test/flutter_test.dart';
import 'package:stack_chan/model/expression_data.dart';

void main() {
  group('ExpressionData', () {
    test('round-trips JSON for avatar payloads', () {
      final expression = ExpressionData(
        leftEye: ExpressionItem(x: 1, y: 2, rotation: 3, weight: 4, size: 5),
        rightEye: ExpressionItem(
          x: 6,
          y: 7,
          rotation: 8,
          weight: 9,
          size: 10,
        ),
        mouth: ExpressionItem(
          x: 11,
          y: 12,
          rotation: 13,
          weight: 14,
          size: 15,
        ),
      );

      final decoded = ExpressionData.fromJson(expression.toJson());

      expect(decoded.type, 'bleAvatar');
      expect(decoded.leftEye.toJson(), expression.leftEye.toJson());
      expect(decoded.rightEye.toJson(), expression.rightEye.toJson());
      expect(decoded.mouth.toJson(), expression.mouth.toJson());
      expect(expression.toString(), contains('"type":"bleAvatar"'));
    });

    test('copy returns a detached expression item', () {
      final original = ExpressionItem(x: 4, y: 5, rotation: 6, weight: 7, size: 8);
      final copy = original.copy();

      copy.x = 99;

      expect(original.x, 4);
      expect(copy.toJson(), isNot(original.toJson()));
      expect(copy.rotation, original.rotation);
      expect(copy.weight, original.weight);
      expect(copy.size, original.size);
    });
  });

  group('MotionData', () {
    test('prefers angle when serializing zero-rotate motions', () {
      final motion = MotionData(
        pitchServo: MotionDataItem(angle: 450, speed: 700),
        yawServo: MotionDataItem(angle: 120, speed: 500),
      );

      final decoded = MotionData.fromJson(motion.toJson());

      expect(decoded.pitchServo.angle, 450);
      expect(decoded.pitchServo.speed, 700);
      expect(decoded.yawServo.toJson(), {'angle': 120, 'speed': 500});
    });

    test('serializes rotate motions without an angle field', () {
      final item = MotionDataItem(rotate: 90, speed: 333);

      expect(item.toJson(), {'rotate': 90, 'speed': 333});
    });
  });

  group('RgbData', () {
    test('uses documented defaults and round-trips JSON', () {
      final rgb = RgbData();
      final decoded = RgbData.fromJson(rgb.toJson());

      expect(decoded.leftRgbColor, '#FFFFFF');
      expect(decoded.rightRgbColor, '#FFFFFF');
      expect(decoded.leftRgbDuration, 0.0);
      expect(decoded.rightRgbDuration, 0.0);
    });
  });
}
